File: src/lib/rng.ts
// Stateless, seedable, hash-based RNG — "random that isn't random". The same (seed, ...keys) ALWAYS
// yields the same value, with NO stream state to advance: you can sample any coordinate directly, in any
// order, without having stepped through the ones before it. That property is the whole point — it lets
// the living world become a PURE FUNCTION of (seed, clock-tick): feed it a point on the clock (see
// clock.ts) and get a deterministic-but-erratic outcome you can reproduce, trace, and predict later. That
// is what makes time travel + shareable evolving worlds possible (see docs/self-sustaining-world.md).
//
// Core hash is Squirrel Eiserloh's noise function (GDC 2017, "Noise-Based RNG") — fast, well-distributed,
// reversible-free. All math is kept in 32-bit lanes via `Math.imul` / `| 0` / `>>> 0`.

const NOISE1 = 0xb5297a4d;
const NOISE2 = 0x68e31da4;
const NOISE3 = 0x1b56c4e9;
const U32 = 2 ** 32;

/** Hash a single 32-bit integer position with a seed → a well-mixed uint32. The atom everything builds on. */
export function hash(position: number, seed = 0): number {
	let m = Math.imul(position | 0, NOISE1);
	m = (m + (seed | 0)) | 0;
	m ^= m >>> 8;
	m = (m + NOISE2) | 0;
	m ^= m << 8;
	m = Math.imul(m, NOISE3);
	m ^= m >>> 8;
	return m >>> 0;
}

/** Fold an arbitrary list of integer keys into one uint32 (order-sensitive; [] ≠ [0] ≠ [0,0]). */
export function hashKeys(seed: number, keys: number[]): number {
	let h = seed | 0;
	for (let i = 0; i < keys.length; i++) h = hash(keys[i], h);
	return hash(keys.length, h); // fold the arity so a trailing zero key can't alias a shorter coord
}

/** A float in [0, 1) at coordinate (seed, ...keys). Keys are treated as 32-bit ints — quantise floats first. */
export function rand(seed: number, ...keys: number[]): number {
	return hashKeys(seed, keys) / U32;
}

/** Map a string (or number) seed → a stable uint32 (FNV-1a) so a world name / share token can seed everything. */
export function seedFrom(s: string | number): number {
	if (typeof s === 'number') return s >>> 0;
	let h = 0x811c9dc5;
	for (let i = 0; i < s.length; i++) {
		h ^= s.charCodeAt(i);
		h = Math.imul(h, 0x01000193);
	}
	return h >>> 0;
}

/** A reusable RNG bound to one seed — ergonomic for the sim/director, and consistent with SimClock's
 *  class style. Every method stays a PURE function of the keys you pass (so it's still fully addressable +
 *  time-travelable); the instance holds only the immutable seed — no mutable stream state. */
export class Rng {
	/** The resolved uint32 seed this RNG is bound to. */
	readonly seed: number;
	// two fixed sub-channels so a Gaussian draws two INDEPENDENT uniforms from the same coordinate
	static readonly #G1 = 0x9e3779b9 | 0;
	static readonly #G2 = 0x85ebca6b | 0;

	constructor(seedInput: string | number) {
		this.seed = seedFrom(seedInput);
	}

	/** [0, 1) at this coordinate. */
	rand(...keys: number[]): number {
		return rand(this.seed, ...keys);
	}
	/** [lo, hi) at this coordinate. */
	range(lo: number, hi: number, ...keys: number[]): number {
		return lo + (hi - lo) * this.rand(...keys);
	}
	/** integer in [loIncl, hiExcl) at this coordinate. */
	int(loIncl: number, hiExcl: number, ...keys: number[]): number {
		return loIncl + Math.floor(this.rand(...keys) * (hiExcl - loIncl));
	}
	/** true with probability p at this coordinate. */
	chance(p: number, ...keys: number[]): boolean {
		return this.rand(...keys) < p;
	}
	/** deterministic element of arr at this coordinate. */
	pick<T>(arr: readonly T[], ...keys: number[]): T {
		return arr[Math.floor(this.rand(...keys) * arr.length)];
	}
	/** A signed unit drift in (-1, 1), handy for "nudge a value a little" at this coordinate. */
	signed(...keys: number[]): number {
		return this.rand(...keys) * 2 - 1;
	}
	/** Gaussian (Box–Muller) with the given mean/std — for trait mutation drift, etc. */
	normal(mean: number, std: number, ...keys: number[]): number {
		const u1 = Math.max(1e-12, this.rand(...keys, Rng.#G1));
		const u2 = this.rand(...keys, Rng.#G2);
		return mean + std * Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
	}
	/** A STATEFUL sub-stream anchored at these keys (folds an auto-incrementing counter) — for "give me N
	 *  draws here" where you don't need to address each one. Still 100% deterministic + reproducible. */
	stream(...keys: number[]): () => number {
		let i = 0;
		return () => this.rand(...keys, i++);
	}
}

File: crates/worldsim/src/rng.rs
//! Stateless squirrel-noise RNG — bit-exact port of `src/lib/rng.ts`. The same `(seed, keys)` always
//! yields the same value with no stream state, so the world is a pure function of (seed, clock-tick).
//! Core hash is Squirrel Eiserloh's noise function (GDC 2017); all lanes are 32-bit wrapping `u32`,
//! which reproduces the JS `Math.imul` / `| 0` / `>>> 0` arithmetic exactly.

const NOISE1: u32 = 0xb529_7a4d;
const NOISE2: u32 = 0x68e3_1da4;
const NOISE3: u32 = 0x1b56_c4e9;
const FNV_OFFSET: u32 = 0x811c_9dc5;
const FNV_PRIME: u32 = 0x0100_0193;

/// Hash a single 32-bit position with a seed → a well-mixed `u32`. The atom everything builds on.
#[inline]
pub fn hash(position: i32, seed: i32) -> u32 {
    let mut m = (position as u32).wrapping_mul(NOISE1);
    m = m.wrapping_add(seed as u32);
    m ^= m >> 8;
    m = m.wrapping_add(NOISE2);
    m ^= m << 8;
    m = m.wrapping_mul(NOISE3);
    m ^= m >> 8;
    m
}

/// Fold a list of integer keys into one `u32` (order-sensitive; the arity is folded in so a trailing
/// zero key can't alias a shorter coordinate).
pub fn hash_keys(seed: u32, keys: &[i32]) -> u32 {
    let mut h = seed;
    for &k in keys {
        h = hash(k, h as i32);
    }
    hash(keys.len() as i32, h as i32)
}

/// A float in `[0, 1)` at coordinate `(seed, keys)`.
#[inline]
pub fn rand(seed: u32, keys: &[i32]) -> f64 {
    hash_keys(seed, keys) as f64 / 4_294_967_296.0 // / 2^32
}

/// Map a string seed → a stable `u32` (FNV-1a over UTF-16 code units, matching JS `charCodeAt`).
pub fn seed_from(s: &str) -> u32 {
    let mut h = FNV_OFFSET;
    for u in s.encode_utf16() {
        h ^= u as u32;
        h = h.wrapping_mul(FNV_PRIME);
    }
    h
}

#[cfg(test)]
mod tests {
    use super::*;
    // Reference values produced by src/lib/rng.ts (the JS source of truth). These MUST match bit-for-bit:
    // it is exactly what guarantees a world evolves identically in JS, the WASM worker, and the CF server tick.
    #[test]
    fn hash_parity() {
        assert_eq!(hash(123, 0), 492_293_964);
        assert_eq!(hash(1, 1), 3_586_333_332);
        assert_eq!(hash(-1, 100), 1_076_820_394); // negative position exercises the u32 wrap
    }
    #[test]
    fn hash_keys_parity() {
        assert_eq!(hash_keys(42, &[1, 2, 3]), 2_609_341_440);
        assert_eq!(hash_keys(7, &[]), 3_589_541_006);
        assert_eq!(hash_keys(0, &[5, 5]), 3_162_176_880);
    }
    #[test]
    fn seed_from_parity() {
        assert_eq!(seed_from("worldgen-agents"), 4_204_040_608); // the live sim's base seed
        assert_eq!(seed_from("hello"), 1_335_831_723);
    }
    #[test]
    fn rand_matches_hash_keys() {
        assert!((0.0..1.0).contains(&rand(42, &[1, 2, 3])));
        assert_eq!(rand(42, &[1, 2, 3]), 2_609_341_440.0 / 4_294_967_296.0);
    }
}
